git-debuggability: paginate recorded commands

BolajiOlajide opened the pull request:
[Walkthrough](https://www.loom.com/share/8278ed0143af4f3d8734a69dafce91b8)

This PR introduces a slice connection resolver that can be used to paginate slices. I added this under the `graphqlutil` package because it fulfills an interface similar to that of the `ConnectionResolver`.

```graphql
type SliceConnectionResolver[T, any] interface {
	Nodes(ctx context.Context) []T
	TotalCount(ctx context.Context) int32
	PageInfo(ctx context.Context) *PageInfo
}
```

This ensures we can adequately paginate the `RecordedCommands` field on the `RepositoryResolver`.

This is particularly useful because we get a slice of byte slice from redis and need to transform that into `RecordedCommandResolver`.

## Test plan

<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles -->

* Navigate to the API console
* Run a GraphQL query to fetch the recorded commands for any repository of your choosing.
```graphql
query {
  repository(name:  <REPOSITORY_NAME>) {
    id
    name
    mirrorInfo {
      shard
    }
    recordedCommands(limit: 100) {
      totalCount
      pageInfo {
        hasNextPage
      }
      nodes {
        dir
        command
        duration
        start
      }
    }
  }
}
```

* You should be able to retrieve the nodes, pageInfo and totalCount of the recorded commands.

eseliger commented on cmd/frontend/graphqlbackend/graphqlutil/slice_connection_resolver_test.go:
```diff
@@ -0,0 +1,76 @@
+		t.Fatal(err)
+	}
+
+	if len(nodes) != 2 {
```
can we get test cases for

- limit out of bounds
- offset out of bounds
- limit 0, offset not 0

please? :)

sashaostrikov replied:
Let's also add a test for nil `transformer` which can totally happen

BolajiOlajide replied:
SUre thing. I'll add those.

sashaostrikov commented on cmd/frontend/graphqlbackend/graphqlutil/slice_connection_resolver.go:
```diff
@@ -0,0 +1,86 @@
+
+type SliceConnectionTransformer[T, V any] func(item T) (V, error)
+
+func NewSliceConnectionResolver[T, V any](data []T, limit, offset int, transformer SliceConnectionTransformer[T, V]) SliceConnectionResolver[T, V] {
```
Agree with Erik, this resolver should be used for pagination as a "last resort" when we cannot apply pagination up the stack.

BolajiOlajide replied:
On second glance. I see the `FIFOList` provides a `Slice` method we can use to paginate from the database itself. Using that instead of `store.All` will enable us to simplify this resolver so it only handles returning data in a GraphQL-compliant way instead  of also handling pagination.

indradhanush commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -3893,7 +3893,36 @@ type Repository implements Node & GenericSearchResultInterface {
     Returns a list of all recorded commands for given repository.
     """
-    recordedCommands: [RecordedCommand!]!
+    recordedCommands(
```
Is it a good idea to add metadata as top level attributes on the `Repository`
node? I'm thinking if we want to namespace this under a `RepositoryMetadata`
type or similar and that leaves the door open to add more stuff. What I'm trying
to avoid is have five different types of metadata attributes on the top level
`Repository` node.

indradhanush replied:
Hmm, I think I skipped a part in my comment. What I mean is to not do something like:

`Repository -> recordedCommands`

But:

`Repository -> RepositoryMetadata -> recordedCommands`

That way future metadata stuff can be added under `RepositoryMetadata` instead
of another edge directly on the `Repository` node.   

Command history per repo makes sense from an API design perspective. 👍

sashaostrikov replied:
Ah, gotcha. I think this is something other than the metadata, TBH. It's more about which commands ran for this repo vs. what are the attributes (metadata) of this repo.

indradhanush replied:
Wouldn't that be metadata? Like history metadata then? It's not direct repository level info though is it? 😁

indradhanush replied:
To clarify I didn't mean that `RepositoryMetadata` should be the name - but just suggested that as a reference. We probably should find a better name since it does feel like it is not the best name for describing based on our discussion already. 😛

BolajiOlajide replied:
I agree with Indra, I think that sounds good to me.

sashaostrikov replied:
I'd postpone the refactor for a bit until we understand that we have anything to add to this category except recorded commands. Otherwise, we will have just another level of putting stuff inside stuff :)

indradhanush replied:
I wouldn't call it a refactor if its API design though. But if you think you can review the design soon-ish (ideally with a tracked issue) it's not a blocker from me. But in my experience it's better to take a step back and figure out the right API design early on instead of going back and reworking stuff because it almost doesn't happen half the time and the other half it takes far too longer. Upto you both on how you want to go ahead. 👍

BolajiOlajide replied:
We'll discuss that and create a separate PR to implement that.

indradhanush commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -3893,7 +3893,36 @@
+    A list of recorded commands for given repository.
+    """
+    nodes: [RecordedCommand!]!
+
```
Is it not possible to not have 0 nodes here? 🤔

sashaostrikov replied:
It is, it's impossible to have a null array (empty is fine) or null entries (zero is fine) 😁

indradhanush replied:
Got it. Wanted to be sure. Thanks for clarifying. 👍

BolajiOlajide replied:
Yup. The inner-most `!` means if the array isn't empty, then it must be a `RecordedCommand, it can't be a null value.

indradhanush commented on cmd/frontend/graphqlbackend/graphqlutil/slice_connection_resolver.go:
```diff
@@ -0,0 +1,86 @@
+	}
+}
+
+type sliceConnectionResolver[T, V any] struct {
```
What is a sliceConnectionResolver? Just looking at this struct it is not
immediately obvious to me which resource this is related to. Maybe some docs
here helps with that? 

A slightly modified version of the PR description intro would be a great fit here.

indradhanush commented on cmd/frontend/graphqlbackend/graphqlutil/slice_connection_resolver.go:
```diff
@@ -0,0 +1,86 @@
+}
+
+type sliceConnectionResolver[T, V any] struct {
+	once sync.Once
```
Maybe this should be named `computeOnce` so that its immediately obvious what this is guarding?

sashaostrikov commented on cmd/frontend/graphqlbackend/graphqlutil/slice_connection_resolver.go:
```diff
@@ -0,0 +1,86 @@
+
+		c.paginatedData = make([]V, len(rawPaginatedData))
+		for i, raw := range rawPaginatedData {
+			c.paginatedData[i], c.err = c.transformer(raw)
```
We need a nil-check for transformer function, I believe

sashaostrikov commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -3893,7 +3893,36 @@ type Repository implements Node & GenericSearchResultInterface {
-    recordedCommands: [RecordedCommand!]!
+    recordedCommands(
+        """
+        Returns the "limit" number recorded commands from the list.
```
maybe like this?
```suggestion
        Number of recorded commands returned.
```

I can see where you took this value from, we can edit it there as well 😁

sashaostrikov commented on cmd/frontend/graphqlbackend/recorded_commands.go:
```diff
@@ -4,38 +4,54 @@ import (
-		return []RecordedCommandResolver{}, nil
+		return graphqlutil.NewSliceConnectionResolver([][]byte{}, 0, 0, recordedCommandTransformer), nil
 	}
 	raws, err := store.All(ctx)
```
`store.Slice()` seems like a good candidate

sashaostrikov commented on cmd/frontend/graphqlbackend/recorded_commands.go:
```diff
@@ -4,38 +4,54 @@ import (
-		}
-		resolvers[i] = NewRecordedCommandResolver(command)
+
+	if args.Limit == nil || *args.Limit > recordedCommandDefaultLimit {
```
Hmm, so we cap at a default value here, but why? Is it the maximum number of entries per page that we want to keep?

BolajiOlajide replied:
We currently allow up to 10K commands to be recorded; nothing stops anyone from making a GraphQL request for 10K at a time which Is a lot.

And yes, this default is the maximum number of entries per page.

BolajiOlajide commented on cmd/frontend/graphqlbackend/recorded_commands.go:
```diff
@@ -4,38 +4,54 @@ import (
 )
 
-func (r *RepositoryResolver) RecordedCommands(ctx context.Context) ([]RecordedCommandResolver, error) {
+const recordedCommandDefaultLimit = 40
```
I've moved this into the graphql schema. I left the const definition here to gate the limit.

BolajiOlajide commented:
@sourcegraph/source I overcomplicated my initial solution here; I've simplified it further and added a video walkthrough to show it in action.

https://www.loom.com/share/8278ed0143af4f3d8734a69dafce91b8

sashaostrikov commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -3893,7 +3893,38 @@ type Repository implements Node & GenericSearchResultInterface {
+    recordedCommands(
+        """
+        Maximum number of recorded commands returned.
+        (Default: 40)
```
Let's add a note that if someone sets the limit to >40, it will be reset to 40, so there are no surprises.

BolajiOlajide replied:
Good catch! I'll do that.

sashaostrikov commented on cmd/frontend/graphqlbackend/graphqlutil/slice_connection_resolver.go:
```diff
@@ -0,0 +1,57 @@
+}
+
+func (c *sliceConnectionResolver[T]) PageInfo(ctx context.Context) *PageInfo {
+	var hasNextPage bool
+	if c.total > c.currentEnd {
+		hasNextPage = true
+	}
+	return HasNextPage(hasNextPage)
```
nit: we can simplify it:

```suggestion
	return HasNextPage(c.total > c.currentEnd)
```

sashaostrikov commented on cmd/frontend/graphqlbackend/graphqlutil/slice_connection_resolver_test.go:
```diff
@@ -0,0 +1,85 @@
+		[]byte("c"),
+	}
+
+	t.Run("paginated slice", func(t *testing.T) {
```
nit: both tests look similar, I guess you can remove duplicated code if you use a table-driven approach, like this: https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/gqlutil/datetime_test.go?L33-51

BolajiOlajide replied:
I agree there's a bunch of duplicated code, but I'll move them into a function call if need be.

A table-driven approach, in this case, makes the tests harder to read, especially because the indentation is off when comparing results.

![CleanShot 2023-07-27 at 19 18 21](https://github.com/sourcegraph/sourcegraph/assets/25608335/34c9efc4-bcb3-4797-b203-ac9e19dc25dd)

sashaostrikov commented on cmd/frontend/graphqlbackend/recorded_commands.go:
```diff
@@ -4,29 +4,57 @@ import (
+	}
+	currentEnd := offset + limit
+
 	recordingConf := conf.Get().SiteConfig().GitRecorder
```
I guess we can make this check in the first line of this function and then return an "empty" resolver, without checking the `offset` and `limit`?

BolajiOlajide replied:
I need the `currentEnd` variable to pass as an argument to the `NewSliceConnectionResolver` function.

sashaostrikov commented on cmd/frontend/graphqlbackend/recorded_commands_test.go:
```diff
@@ -35,170 +35,312 @@ func TestRecordedCommandsResolver(t *testing.T) {
+										dir
+										path
+									}
+									totalCount
```
Let's also check `hasNextPage` in all tests.

sashaostrikov commented on cmd/frontend/graphqlbackend/recorded_commands_test.go:
```diff
@@ -35,170 +35,312 @@ func TestRecordedCommandsResolver(t *testing.T) {
+				Query: `
+						{
+							repository(name: "github.com/sourcegraph/sourcegraph") {
+								recordedCommands(limit: 10000) {
```
If we could mock the default limit of 40 in this test (setting it to 2, e.g.), then it would also test the resetting, but in case of limit=40 and 3 total entries, we cannot 100% tell that it is tested.

It's a simple if statement to test, but still :)

BolajiOlajide replied:
Added this. Thanks.

sashaostrikov commented on cmd/frontend/graphqlbackend/recorded_commands_test.go:
```diff
@@ -35,170 +35,312 @@ func TestRecordedCommandsResolver(t *testing.T) {
+						}
+					`,
+			})
+		})
```
Can you pls add a test for both valid `offset` and `limit`?

sashaostrikov reviewed (commented):
LGTM, didn't hit "approve" right away because asked for one more test, nothing critical apart from that, thanks for adding a bunch of tests 👍🏻

sashaostrikov reviewed (approved):
LGTM!
